modern/src/Actions.ts

```diff
@@ -34,8 +35,24 @@ export function gotSkinBlob(blob: Blob, store: ModernStore) {
   };
 }
 
+async function unloadSkin(makiTree) {
+  await Utils.asyncTreeFlatMap(makiTree, async (node: MakiObject) => {
+    if (node instanceof JsScript && node.system) {
+      node.system.onscriptunloading();
+    }
+
+    return node;
+  });
+}
+
 export function gotSkinZip(zip: JSZip, store: ModernStore) {
   return async dispatch => {
+    // unload current skin if one has been loaded
+    if (store.getState().modernSkin.skinLoaded) {
+      unloadSkin(store.getState().modernSkin.makiTree);
+      dispatch({ type: "UNLOAD_SKIN" });
```

Comment:
Can we `await unloadSkin()` and then dispatch an action in the past tense: `"SKIN_UNLOADED"`.

Reply:
Oh whoops, definitely meant to `await` that call, thanks!

Yup, I'll change the name.